### DISCUSSION TITLE
How do I get copy on selection to work when I cmd+tab out?

### DISCUSSION THREAD

**Discussion by aliceisjustplaying** [wezterm/wezterm#6408] [Q&A]
### What Operating System(s) are you running on?

macOS

### WezTerm version

20240915-153243-2d0c5cdd

### Ask your question!

I have Wezterm set up for copy on selection, but crucially, if I:
1. select a text
2. without releasing the left mouse button, I cmd+tab to a different window
3. try to paste what I selected

...It fails, the selection was not copied to the clipboard. 

This sucks because this totally works in both iTerm 2 and VSCode and at this point is deep muscle memory. Is there a fix?

My config is below.

```lua
local wezterm = require 'wezterm'
local config = wezterm.config_builder()
local act = wezterm.action

function make_mouse_binding(dir, streak, button, mods, action)
    return {
        event = { [dir] = { streak = streak, button = button } },
        mods = mods,
        action = action,
    }
end

-- For example, changing the color scheme:
config.color_scheme = 'Catppuccin Macchiato'
config.font_size = 16
config.font = wezterm.font('Cascadia Code NF')
config.harfbuzz_features = { 'calt=0', 'clig=0', 'liga=0' }
config.mouse_bindings = {
    make_mouse_binding('Up', 1, 'Left', 'NONE',
        wezterm.action.CompleteSelectionOrOpenLinkAtMouseCursor 'ClipboardAndPrimarySelection'),
    make_mouse_binding('Up', 1, 'Left', 'SHIFT',
        wezterm.action.CompleteSelectionOrOpenLinkAtMouseCursor 'ClipboardAndPrimarySelection'),
    make_mouse_binding('Up', 1, 'Left', 'ALT', wezterm.action.CompleteSelection 'ClipboardAndPrimarySelection'),
    make_mouse_binding('Up', 1, 'Left', 'SHIFT|ALT',
        wezterm.action.CompleteSelectionOrOpenLinkAtMouseCursor 'ClipboardAndPrimarySelection'),
    make_mouse_binding('Up', 2, 'Left', 'NONE', wezterm.action.CompleteSelection 'ClipboardAndPrimarySelection'),
    make_mouse_binding('Up', 3, 'Left', 'NONE', wezterm.action.CompleteSelection 'ClipboardAndPrimarySelection'),
    {
        event = { Down = { streak = 1, button = "Right" } },
        mods = "NONE",
        action = wezterm.action_callback(function(window, pane)
            local has_selection = window:get_selection_text_for_pane(pane) ~= ""
            if has_selection then
                window:perform_action(act.CopyTo("ClipboardAndPrimarySelection"), pane)
                window:perform_action(act.ClearSelection, pane)
            else
                window:perform_action(act({ PasteFrom = "Clipboard" }), pane)
            end
        end),
    }
}
return config
```

**Comment by bew** [comment]
I think instead of doing the copy on `Up` event, you'd have to do the copy on `Drag` event 🤔

**Reply by aliceisjustplaying** [reply]
I've tried changing it to `Drag` but that just breaks it completely

**Reply by bew** [reply]
Hmm probably because they run too early.. or you'd have to trigger both the selection extend AND the copy (but I'm not sure it's properly synchronous)
:point_right: Maybe instead of removing the Up bindings, you could try registering an event handler on the focus change, and copy the selection if there's one :thinking: 

refs:
https://wezfurlong.org/wezterm/config/lua/window-events/window-focus-changed.html
https://wezfurlong.org/wezterm/config/lua/window/get_selection_text_for_pane.html

**Reply by aliceisjustplaying** [reply]
thanks! i *think* this did the job:

```lua
wezterm.on('window-focus-changed', function(window, pane)
    local has_selection = window:get_selection_text_for_pane(pane) ~= ""
    if has_selection then
        window:perform_action(act.CopyTo("ClipboardAndPrimarySelection"), pane)
    end
end)
```

I continue to be baffled how something as core as this is not a simple option, but oh well

**Reply by bew** [reply]
> I continue to be baffled how something as core as this is not a simple option, but oh well

Well, it seems no-one was ever bummed by this to raise an issue, I never used copying like this personally 🤷‍♂️

As for the default, @wez it might be a good idea indeed to run mouse `Up` event if a drag is in progress and window focus is changing 🤔 WDYT?

**Reply by wez** [reply]
> might be a good idea indeed to run mouse Up event if a drag is in progress and window focus is changing 🤔 WDYT?

The events reflect the events received by wezterm, so if we didn't get an Up event, it feels maybe a little weird to synthesize one, although I think we do something similar when generating the reports we send into the terminal.

> I continue to be baffled how something as core as this is not a simple option, but oh well

You know, comments like that immediately de-prioritize any work that I might have felt inspired to do here.
It is possible to ask for something to be more streamlined *without* coming across as an asshole.

**Reply by aliceisjustplaying** [reply]
I was frustrated but yes, that was a bit much and I apologize.

**Reply by aliceisjustplaying** [reply]
Right, that's still an excuse. I apologize, period.
